### ISSUE TITLE
torch.mul does not reshape its output gradient

### ISSUE THREAD

**Issue by bpiwowar** [pytorch/pytorch#818]
I think that there is a problem with torch.mul that does not reshape correctly its output gradient. In the code below, the first backward works, but the second does not.

```
import torch
from torch.autograd import Variable
from torch.nn import Linear

a = Variable(torch.randn(5, 12), requires_grad=True)
b = Variable(torch.randn(5), requires_grad=True)

l = torch.nn.Linear(12, 1)

# Works
c = (l.forward(a) * b).sum()
c.backward()

# RuntimeError: matrices expected, got 1D, 2D tensors at .../torch/lib/TH/generic/THTensorMath.c:857
c = (b * l.forward(a)).sum()
c.backward()
```

**Comment by apaszke** [comment]
Thanks for the report! Just note that you shouldn't use the forward method of the module, but call it like a function instead. It won't fix the bug, but it's quite important, because the hooks won't work properly without that.

**Comment by apaszke** [comment]
Fixed in #852.
